source/common/network/io_socket_handle_impl.cc

```diff
@@ -95,6 +95,13 @@ Api::IoCallUint64Result IoSocketHandleImpl::readv(uint64_t max_length, Buffer::R
     num_bytes_to_read += slice_length;
   }
   ASSERT(num_bytes_to_read <= max_length);
+
+  if (num_slices_to_read == 1) {
+    // Avoid paying the VFS overhead when there is only one IO buffer to work with
+    return sysCallResultToIoCallResult(
+      Api::OsSysCallsSingleton::get().recv(fd_, iov[0].iov_base, iov[0].iov_len, 0));
```

Comment:
emm...if this give good performance, I'm little to prefer to replace writev/readv with send/recv at here
https://github.com/envoyproxy/envoy/blob/0086411dd85231434c28ffb6af0ebeaf49e6b02f/source/extensions/transport_sockets/tls/io_handle_bio.cc#L74

But yea, that means we need to figure out what to do for mac and windows. If we just want to focus on linux platform, then this is probably the right place.

Reply:
Agree that we can change https://github.com/envoyproxy/envoy/blob/0086411dd85231434c28ffb6af0ebeaf49e6b02f/source/extensions/transport_sockets/tls/io_handle_bio.cc#L74 to use send() since it explicitly specify the number of slices to be 1.

I wonder whether it makes sense to generalize the optimization (not only TLS but other filters like Redis) - this PR tries to look at the number of slices in run time and calls send()/recv() if the number of slices is 1.